### ISSUE TITLE
Please add checkpointing and saving of weights

### ISSUE THREAD

**Issue by azw413** [ssoudan/nanogptrs#8]
Thanks for the great work. I ran it last night (after a few changes to get it working with `tch = "0.13.0"`) and it panicked after training :( Consequently, I couldn't actually run the model unless I wait another 7 hours for training. Maybe you could you add checkpointing to save the weights after each epoch and then on the commandline allow a weight file to be used either as a basis for fine tuning or for inference? It also might be good to make the lib.rs a crate which can be included in other projects and move main.rs to an example. Great work though!

**Comment by ssoudan** [comment]
Fixed the panic in the latest commit. 

For the Save/Load, it makes sense. If you want to have a shot at it, I'm happy to provide some guidance if needed. Here is a start: The VarStore (https://docs.rs/tch/latest/tch/nn/struct.VarStore.html) has the methods to do this. Clap subcommands could be a way to distinguish between the `train` and `generate` workflow (see https://github.com/clap-rs/clap/blob/master/examples/tutorial_derive/03_04_subcommands_alt.rs).

As for making a library of it. I'm not opposed getting a MR going in this direction but this is a 'toy' model and not sure it's worth the effort.

**Comment by ssoudan** [comment]
Also `cargo run --release -- --device=cuda --n-epochs=1 nano-gpt -H 3 -L 3` is much faster (~45min) and gives a result where its visible the LM learnt something.

**Comment by azw413** [comment]
Thanks for the tips, I added saving and loading the VarStore and this was very easy and worked really well. Karpathy mentions loading GPT-2 weights directly which would save a huge amount of training. I guess this would depend on having an identical network configuration and using the exact same tokenizer as GPT-2. I might try to add the BPE tokenizer from HuggingFace's `tokenizers` Rust crate to the simple Shakespeare demo. Any thoughts on loading the GPT-2 weights? I'm assuming it's the 'small' 124M parameter version of GPT-2. What NanoGPT config would equate to that? Thanks.

**Comment by ssoudan** [comment]
Here is an example of loading a tokenizer from the json file: https://github.com/ssoudan/sapiens/blob/a6639bd1fc913081d653662be98d15672be6d140/sapiens/src/models/openai.rs#L88

--
Once you get the structure correct - number of heads, layers, size of the embeddings...
You will have to align the names of the tensors in the model and in the weight file.

You can list them from the VarStore with something like: https://github.com/ssoudan/nanogptrs/blob/9053474a4625f8383a680ee24e785ec5f11f13be/src/main.rs#L247.

Then to align them, not sure which of the following route is easier: 
- changing the names ('Path') in the src/model.rs as the modules are constructed
- or; figuring out the mapping and copying the weights from one VarStore to another
- or; creating a separate tool to convert the weight file: https://github.com/LaurentMazare/tch-rs/blob/59fb16165eb9cdcbe788f91657f67fde37cb67fd/examples/tensor-tools.rs#L13

**Comment by azw413** [comment]
Here's a good example from Karpathy of what I'll need to do:- https://github.com/karpathy/nanoGPT/blob/master/model.py#L249 
Thanks for the tip on using the tokenizer, I'm using the GPT2 tokenizer from the HuggingFace tokenizers crate which seems to be working. One thing I've noticed though is that the VRAM usage is huge, e.g. 27GB for a GPT2 sized model and even 19GB on the default 4 x 4 layer model. I noticed that the input and output tensors are i64. In the Python code, I couldn't see what the types are but maybe there's some scope for optimising the memory usage there.

**Comment by ssoudan** [comment]
27GB seems reasonable for a 'small' GPT-2 trained in f32: https://www.trentonbricken.com/TransformerMemoryRequirements/

PyTorch NLL loss requires i64/long types. 
[The NLL](https://github.com/pytorch/pytorch/issues/45540)

The weights are in f32. You can see that be enumerating the variables - with their types on the VarStore.

**Comment by ssoudan** [comment]
Added in #10
